File: src/encoding/strict-json.ts
/**
 * Strict JSON parser for Sunlight wire/protocol documents (spec §2).
 *
 *  - UTF-8 input; BOM rejected; trailing non-whitespace bytes rejected.
 *  - Duplicate object keys rejected before object construction.
 *  - Lone surrogates rejected; no replacement-character repair.
 *  - Numbers are integers in [0, 9007199254740991] only: raw `-0`, negative
 *    numbers, fractional notation, and exponent notation are rejected at the
 *    lexical level before any value is constructed.
 *  - Nesting deeper than 32 levels rejected.
 */

import { SunlightError } from "../errors.js";

export const MAX_JSON_DEPTH = 32;
export const MAX_SAFE_UINT = 9007199254740991;

export class StrictJsonError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "StrictJsonError";
  }
}

export type JsonValue =
  | null
  | boolean
  | number
  | string
  | JsonValue[]
  | { [key: string]: JsonValue };

function hasLoneSurrogate(s: string): boolean {
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c >= 0xd800 && c <= 0xdbff) {
      const n = s.charCodeAt(i + 1);
      if (!(n >= 0xdc00 && n <= 0xdfff)) return true;
      i++;
    } else if (c >= 0xdc00 && c <= 0xdfff) {
      return true;
    }
  }
  return false;
}

class Parser {
  private pos = 0;
  constructor(private readonly src: string) {}

  private peek(): string {
    return this.src[this.pos] ?? "";
  }

  private fail(msg: string): never {
    throw new StrictJsonError(`${msg} at offset ${this.pos}`);
  }

  private expect(ch: string): void {
    if (this.src[this.pos] !== ch) this.fail(`expected ${JSON.stringify(ch)}`);
    this.pos++;
  }

  private skipWs(): void {
    while (this.pos < this.src.length) {
      const c = this.src.charCodeAt(this.pos);
      if (c === 0x20 || c === 0x09 || c === 0x0a || c === 0x0d) this.pos++;
      else break;
    }
  }

  parse(): JsonValue {
    this.skipWs();
    const v = this.value(0);
    this.skipWs();
    if (this.pos !== this.src.length) this.fail("trailing content");
    return v;
  }

  private value(depth: number): JsonValue {
    if (depth > MAX_JSON_DEPTH) this.fail("depth limit exceeded");
    const c = this.peek();
    switch (c) {
      case "{":
        return this.object(depth);
      case "[":
        return this.array(depth);
      case '"':
        return this.string();
      case "t":
        return this.literal("true", true);
      case "f":
        return this.literal("false", false);
      case "n":
        return this.literal("null", null);
      default:
        if (c === "-" || (c >= "0" && c <= "9")) return this.number();
        this.fail("unexpected character");
    }
  }

  private literal(word: string, val: JsonValue): JsonValue {
    if (this.src.startsWith(word, this.pos)) {
      this.pos += word.length;
      return val;
    }
    this.fail(`invalid literal`);
  }

  private object(depth: number): JsonValue {
    this.expect("{");
    const out: Record<string, JsonValue> = {};
    this.skipWs();
    if (this.peek() === "}") {
      this.pos++;
      return out;
    }
    for (;;) {
      this.skipWs();
      if (this.peek() !== '"') this.fail("expected object key");
      const key = this.string() as string;
      if (Object.prototype.hasOwnProperty.call(out, key)) {
        this.fail("duplicate object key");
      }
      this.skipWs();
      this.expect(":");
      this.skipWs();
      out[key] = this.value(depth + 1);
      this.skipWs();
      const c = this.peek();
      if (c === ",") {
        this.pos++;
        continue;
      }
      if (c === "}") {
        this.pos++;
        return out;
      }
      this.fail("expected ',' or '}'");
    }
  }

  private array(depth: number): JsonValue {
    this.expect("[");
    const out: JsonValue[] = [];
    this.skipWs();
    if (this.peek() === "]") {
      this.pos++;
      return out;
    }
    for (;;) {
      this.skipWs();
      out.push(this.value(depth + 1));
      this.skipWs();
      const c = this.peek();
      if (c === ",") {
        this.pos++;
        continue;
      }
      if (c === "]") {
        this.pos++;
        return out;
      }
      this.fail("expected ',' or ']'");
    }
  }

  private string(): string {
    this.expect('"');
    const chunks: string[] = [];
    let start = this.pos;
    while (this.pos < this.src.length) {
      const c = this.src.charCodeAt(this.pos);
      if (c === 0x22) {
        chunks.push(this.src.slice(start, this.pos));
        this.pos++;
        const s = chunks.join("");
        if (hasLoneSurrogate(s)) this.fail("lone surrogate in string");
        return s;
      }
      if (c === 0x5c) {
        chunks.push(this.src.slice(start, this.pos));
        this.pos++;
        const e = this.src[this.pos];
        switch (e) {
          case '"':
          case "\\":
          case "/":
            chunks.push(e);
            this.pos++;
            break;
          case "b":
            chunks.push("\b");
            this.pos++;
            break;
          case "f":
            chunks.push("\f");
            this.pos++;
            break;
          case "n":
            chunks.push("\n");
            this.pos++;
            break;
          case "r":
            chunks.push("\r");
            this.pos++;
            break;
          case "t":
            chunks.push("\t");
            this.pos++;
            break;
          case "u": {
            const hex = this.src.slice(this.pos + 1, this.pos + 5);
            if (!/^[0-9a-fA-F]{4}$/.test(hex)) this.fail("bad \\u escape");
            const cp = parseInt(hex, 16);
            if (cp >= 0xd800 && cp <= 0xdbff && this.src[this.pos + 5] === "\\" && this.src[this.pos + 6] === "u") {
              const hex2 = this.src.slice(this.pos + 7, this.pos + 11);
              if (/^[0-9a-fA-F]{4}$/.test(hex2)) {
                const cp2 = parseInt(hex2, 16);
                if (cp2 >= 0xdc00 && cp2 <= 0xdfff) {
                  chunks.push(String.fromCodePoint(0x10000 + ((cp - 0xd800) << 10) + (cp2 - 0xdc00)));
                  this.pos += 11;
                  break;
                }
              }
            }
            if (cp >= 0xd800 && cp <= 0xdfff) this.fail("lone surrogate escape");
            chunks.push(String.fromCharCode(cp));
            this.pos += 5;
            break;
          }
          default:
            this.fail("bad escape");
        }
        start = this.pos;
        continue;
      }
      if (c < 0x20) this.fail("unescaped control character");
      this.pos++;
    }
    this.fail("unterminated string");
  }

  private number(): number {
    // Protocol numbers: `0 | [1-9][0-9]*` only.  Negative sign, fraction, and
    // exponent notation are rejected before any numeric value is formed.
    const re = /^-?\d+(?:\.\d+)?(?:[eE][+-]?\d+)?/;
    const m = re.exec(this.src.slice(this.pos));
    if (!m) this.fail("invalid number");
    const lexeme = m[0];
    this.pos += lexeme.length;
    if (!/^(?:0|[1-9]\d*)$/.test(lexeme)) {
      this.fail("number outside protocol integer domain");
    }
    const v = Number(lexeme);
    if (!Number.isSafeInteger(v) || v > MAX_SAFE_UINT) {
      this.fail("number out of range");
    }
    return v;
  }
}

/** Parse a UTF-8 buffer strictly. Throws SunlightError(SCHEMA_INVALID). */
export function parseStrictJson(buf: Buffer | Uint8Array): JsonValue {
  try {
    if (buf.length >= 3 && buf[0] === 0xef && buf[1] === 0xbb && buf[2] === 0xbf) {
      throw new StrictJsonError("BOM rejected");
    }
    let src: string;
    try {
      src = new TextDecoder("utf-8", { fatal: true }).decode(buf);
    } catch {
      throw new StrictJsonError("invalid UTF-8");
    }
    return new Parser(src).parse();
  } catch (e) {
    if (e instanceof StrictJsonError) {
      throw new SunlightError("SCHEMA_INVALID", null, e);
    }
    throw e;
  }
}

/** Parse a JS string strictly (for non-byte sources such as fixtures). */
export function parseStrictJsonString(src: string): JsonValue {
  return parseStrictJson(Buffer.from(src, "utf8"));
}

File: python/sunlight/strictjson.py
"""Strict JSON parser for Sunlight wire/protocol documents (spec §2).

- UTF-8 input; BOM rejected; trailing non-whitespace rejected.
- Duplicate object keys rejected before object construction.
- Lone surrogates rejected; no replacement-character repair.
- Numbers are integers in [0, 9007199254740991] only: raw ``-0``, negatives,
  fractions, and exponents are rejected at the lexical level.
- Nesting deeper than 32 levels rejected.

All failures raise ``SunlightError("SCHEMA_INVALID")``.
"""

from __future__ import annotations

from .errors import SunlightError

MAX_JSON_DEPTH = 32
MAX_SAFE_UINT = 9007199254740991

_WS = " \t\n\r"
_ESCAPES = {'"': '"', "\\": "\\", "/": "/", "b": "\b", "f": "\f", "n": "\n", "r": "\r", "t": "\t"}


def _fail(msg: str) -> None:
    raise SunlightError("SCHEMA_INVALID", cause=ValueError(msg))


class _Parser:
    def __init__(self, src: str):
        self.s = src
        self.i = 0

    def parse(self):
        v = self.value(0)
        while self.i < len(self.s) and self.s[self.i] in _WS:
            self.i += 1
        if self.i != len(self.s):
            _fail("trailing bytes")
        return v

    def value(self, depth: int):
        if depth > MAX_JSON_DEPTH:
            _fail("depth limit")
        self._ws()
        if self.i >= len(self.s):
            _fail("unexpected end")
        ch = self.s[self.i]
        if ch == "{":
            return self.obj(depth)
        if ch == "[":
            return self.arr(depth)
        if ch == '"':
            return self.string()
        if ch == "t" and self.s.startswith("true", self.i):
            self.i += 4
            return True
        if ch == "f" and self.s.startswith("false", self.i):
            self.i += 5
            return False
        if ch == "n" and self.s.startswith("null", self.i):
            self.i += 4
            return None
        if ch == "-" or ch.isdigit():
            return self.number()
        _fail(f"unexpected character {ch!r}")

    def _ws(self) -> None:
        while self.i < len(self.s) and self.s[self.i] in _WS:
            self.i += 1

    def obj(self, depth: int):
        self.i += 1
        out = {}
        self._ws()
        if self.i < len(self.s) and self.s[self.i] == "}":
            self.i += 1
            return out
        while True:
            self._ws()
            if self.i >= len(self.s) or self.s[self.i] != '"':
                _fail("expected object key")
            k = self.string()
            if k in out:
                _fail("duplicate object key")
            self._ws()
            if self.i >= len(self.s) or self.s[self.i] != ":":
                _fail("expected ':'")
            self.i += 1
            out[k] = self.value(depth + 1)
            self._ws()
            if self.i >= len(self.s):
                _fail("unterminated object")
            ch = self.s[self.i]
            if ch == ",":
                self.i += 1
                continue
            if ch == "}":
                self.i += 1
                return out
            _fail("expected ',' or '}'")

    def arr(self, depth: int):
        self.i += 1
        out = []
        self._ws()
        if self.i < len(self.s) and self.s[self.i] == "]":
            self.i += 1
            return out
        while True:
            out.append(self.value(depth + 1))
            self._ws()
            if self.i >= len(self.s):
                _fail("unterminated array")
            ch = self.s[self.i]
            if ch == ",":
                self.i += 1
                continue
            if ch == "]":
                self.i += 1
                return out
            _fail("expected ',' or ']'")

    def string(self) -> str:
        self.i += 1
        out = []
        while True:
            if self.i >= len(self.s):
                _fail("unterminated string")
            ch = self.s[self.i]
            if ch == '"':
                self.i += 1
                return "".join(out)
            if ch == "\\":
                self.i += 1
                if self.i >= len(self.s):
                    _fail("unterminated escape")
                e = self.s[self.i]
                if e == "u":
                    hexs = self.s[self.i + 1 : self.i + 5]
                    if len(hexs) != 4 or any(c not in "0123456789abcdefABCDEF" for c in hexs):
                        _fail("bad \\u escape")
                    cp = int(hexs, 16)
                    self.i += 5
                    if 0xD800 <= cp <= 0xDBFF:
                        if self.s[self.i : self.i + 2] == "\\u":
                            lo = int(self.s[self.i + 2 : self.i + 6], 16)
                            if 0xDC00 <= lo <= 0xDFFF:
                                self.i += 6
                                cp = 0x10000 + ((cp - 0xD800) << 10) + (lo - 0xDC00)
                            else:
                                _fail("lone surrogate escape")
                        else:
                            _fail("lone surrogate escape")
                    elif 0xDC00 <= cp <= 0xDFFF:
                        _fail("lone surrogate escape")
                    out.append(chr(cp))
                    continue
                if e not in _ESCAPES:
                    _fail("bad escape")
                out.append(_ESCAPES[e])
                self.i += 1
                continue
            if ord(ch) < 0x20:
                _fail("control character in string")
            if 0xD800 <= ord(ch) <= 0xDFFF:
                _fail("lone surrogate")
            out.append(ch)
            self.i += 1

    def number(self):
        start = self.i
        if self.s[self.i] == "-":
            _fail("negative numbers outside protocol integer domain")
        while self.i < len(self.s) and self.s[self.i].isdigit():
            self.i += 1
        lexeme = self.s[start : self.i]
        if self.i < len(self.s) and self.s[self.i] in ".eE":
            _fail("non-integer number")
        if lexeme == "":
            _fail("empty number")
        if len(lexeme) > 1 and lexeme[0] == "0":
            _fail("leading zero")
        v = int(lexeme)
        if v > MAX_SAFE_UINT:
            _fail("number out of range")
        return v


def parse_strict_json(data) -> object:
    """Parse UTF-8 bytes strictly; ``SunlightError(SCHEMA_INVALID)`` on any failure."""
    if isinstance(data, str):
        data = data.encode("utf-8")
    if data[:3] == b"\xef\xbb\xbf":
        _fail("BOM rejected")
    try:
        src = bytes(data).decode("utf-8", errors="strict")
    except UnicodeDecodeError as e:
        raise SunlightError("SCHEMA_INVALID", cause=e) from e
    return _Parser(src).parse()


def parse_strict_json_str(src: str):
    return parse_strict_json(src.encode("utf-8"))
